processing/src/main/java/io/druid/query/cache/CacheKeyBuilder.java

```diff
@@ -0,0 +1,251 @@
+import java.util.Iterator;
+import java.util.List;
+
+public class CacheKeyBuilder
+{
+  static final byte BYTE_KEY = 0;
+  static final byte BYTE_ARRAY_KEY = 1;
+  static final byte BOOLEAN_KEY = 2;
+  static final byte INT_KEY = 3;
+  static final byte FLOAT_KEY = 4;
+  static final byte FLOAT_ARRAY_KEY = 5;
+  static final byte DOUBLE_KEY = 6;
+  static final byte STRING_KEY = 7;
+  static final byte STRING_LIST_KEY = 8;
+  static final byte CACHEABLE_KEY = 9;
+  static final byte CACHEABLE_LIST_KEY = 10;
+
+  static final byte[] STRING_SEPARATOR = new byte[]{(byte) 0xFF};
+  static final byte[] EMPTY_BYTES = StringUtils.EMPTY_BYTES;
+
+  private static class Item
+  {
+    private final byte typeKey;
+    private final byte[] item;
+
+    Item(byte typeKey, byte[] item)
+    {
+      this.typeKey = typeKey;
+      this.item = item;
+    }
+
+    int byteSize()
+    {
+      return 1 + item.length;
+    }
+  }
+
+  private static byte[] floatArrayToByteArray(float[] input)
+  {
+    final ByteBuffer buffer = ByteBuffer.allocate(Floats.BYTES * input.length);
+    buffer.asFloatBuffer().put(input);
+    return buffer.array();
+  }
+
+  private static byte[] cacheableToByteArray(@Nullable Cacheable cacheable)
+  {
+    if (cacheable == null) {
+      return EMPTY_BYTES;
+    } else {
+      final byte[] key = cacheable.getCacheKey();
+      Preconditions.checkArgument(!Arrays.equals(key, EMPTY_BYTES), "cache key is equal to the empty key");
+      return key;
+    }
+  }
+
+  private static byte[] stringCollectionToByteArray(Collection<String> input)
+  {
+    return collectionToByteArray(
+        input,
+        new Function<String, byte[]>()
+        {
+          @Override
+          public byte[] apply(@Nullable String input)
+          {
+            return StringUtils.toUtf8WithNullToEmpty(input);
+          }
+        },
+        STRING_SEPARATOR
+    );
+  }
+
+  private static byte[] cacheableCollectionToByteArray(Collection<? extends Cacheable> input)
+  {
+    return collectionToByteArray(
+        input,
+        new Function<Cacheable, byte[]>()
+        {
+          @Override
+          public byte[] apply(@Nullable Cacheable input)
+          {
+            return input == null ? EMPTY_BYTES : input.getCacheKey();
+          }
+        },
+        EMPTY_BYTES
+    );
+  }
+
+  private static <T> byte[] collectionToByteArray(
+      Collection<? extends T> collection,
+      Function<T, byte[]> serializeFunction,
+      byte[] separator
+  )
+  {
+    if (collection.size() > 0) {
+      List<byte[]> byteArrayList = Lists.newArrayListWithCapacity(collection.size());
+      int totalByteLength = 0;
+      for (T eachItem : collection) {
+        final byte[] byteArray = serializeFunction.apply(eachItem);
+        totalByteLength += byteArray.length;
+        byteArrayList.add(byteArray);
+      }
+
+      // Sort the byte array list to guarantee that collections of same items but in different orders make the same result
```

Comment:
Good catch. Thanks. I added appendCacheablesIgnoringOrder(). 
I checked which one of appendCacheables and appendCacheablesIgnoringOrder should be called when building cache keys, but am not sure for SketchSetPostAggregator. Please verify it.